### PR TITLE
On the empty string, `.tail` and `.init` now throw (instead of returning the empty string)

### LINKED ISSUE #13026: Empty string tail and init don't throw (unlike WrappedString)
## Reproduction steps

Scala version: 2.13.14

```scala
scala> "".tail
val res0: String = ""

scala> "".init
val res1: String = ""

scala> def tailOf[A](x: Seq[A]) = x.tail
def tailOf[A](x: Seq[A]): Seq[A]

scala> tailOf("")
java.lang.UnsupportedOperationException
  at scala.collection.IterableOps.tail(Iterable.scala:528)
  at scala.collection.IterableOps.tail$(Iterable.scala:527)
  at scala.collection.AbstractIterable.tail(Iterable.scala:935)
  at tailOf(<console>:1)
  ... 32 elided

scala> def initOf[A](x: Seq[A]) = x.init
def initOf[A](x: Seq[A]): Seq[A]

scala> initOf("")
java.lang.UnsupportedOperationException
  at scala.collection.IterableOps.init(Iterable.scala:536)
  at scala.collection.IterableOps.init$(Iterable.scala:535)
  at scala.collection.AbstractIterable.init(Iterable.scala:935)
  at initOf(<console>:1)
  ... 32 elided
```

## Problem

I guess I expected them to be consistent with other collections and throw when empty.

### PR DESCRIPTION
**Pull request by sh0hei**
fix https://github.com/scala/bug/issues/13026

### CODE REVIEW

File: src/library/scala/collection/StringOps.scala [hunk 1]
```diff
@@ -1201,14 +1201,16 @@ final class StringOps(private val s: String) extends AnyVal {
   def withFilter(p: Char => Boolean): StringOps.WithFilter = new StringOps.WithFilter(p, s)
 
   /** The rest of the string without its first char.
-    * @note $unicodeunaware
-    */
-  def tail: String = slice(1, s.length)
+   *  @throws NoSuchElementException if the string is empty.
+   *  @note $unicodeunaware
+   */
+  def tail: String = if(s.isEmpty) throw new NoSuchElementException("tail of empty String") else slice(1, s.length)
```
Comment (som-snytt) [inline src/library/scala/collection/StringOps.scala]: Most collections follow `IterableOps` in throwing `UnsupportedOperation` (but `NSE` for `head` and `last`).

A couple deviate and throw `NSE`. It's fine to leave it as it is here.

Example showing difficulty of enforcing uniformity of behavior:
```
scala> 0.until(0).tail
java.util.NoSuchElementException: tail on empty Range
  at scala.collection.immutable.Range$.scala$collection$immutable$Range$$emptyRangeError(Range.scala:631)
  at scala.collection.immutable.Range.tail(Range.scala:144)
  ... 30 elided

scala> 0.to(0).tail
val res4: scala.collection.immutable.Range = empty Range 0 until 0
```
example with descriptive message:
```
scala> collection.immutable.ArraySeq(Array(42)).tail.tail
java.lang.UnsupportedOperationException: tail of empty array
  at scala.collection.ArrayOps$.tail$extension(ArrayOps.scala:346)
  at scala.collection.immutable.ArraySeq.tail(ArraySeq.scala:242)
  ... 30 elided
```

Comment (som-snytt) [reply src/library/scala/collection/StringOps.scala]: I meant to add, I'll merge as-is if you don't change it. Merges are paused until 2.13.15 is released, I assume.

**Review by som-snytt — APPROVED** [review]

**Review by Philippus — COMMENTED** [review]
most "of empty" exceptions lowercase the type, e.g. "tail of empty array" / "tail of empty list". Would be nice to keep that consistent and use "tail of empty string" / "init of empty string".

**Comment by som-snytt** [comment]
Nice observation. Exception as shown is `empty Range`. If I manage to follow up with a uniform behavior test, I'll include that condition.

I always have a hiccup on String vs string, and favor String FSR. I would probably prefer naming a type, `empty Map`, but there is an argument to be made for either choice. I am not above: "I need a kleenex."

**Comment by SethTisue** [comment]
It is rather bewildering trying to figure out what the "right" or "best" exception to throw is here, but I'll go along with the `UnsupportedOperationException` suggestion.

I have taken the liberty of squashing and force-pushing, so we can hit merge as soon as CI likes it.

I will release-note this since it wouldn't surprise me if there is code out there inadvertently relying on the old behavior.

**Review by SethTisue — APPROVED** [review]

**Review by som-snytt — COMMENTED** [review]
I merged on Seth's approval, but normally I would have complained that there is no space after `if`.

**Comment by Mizommz** [comment]
#10851

**Review by Mizommz — APPROVED** [review]

**Review by Mizommz — APPROVED** [review]

**Pull request merged**